### ISSUE TITLE
Hall of Fame during Board Game coevolution

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEATv2#395]
Look up papers that apply co-evolution to board games using a hall of fame. A hall of fame is a collection of evolved agents across several generations that were the best of their given generation. Halls of fame are typically used with coevolution as a way to assure that learned skills are not forgotten. The population keeps getting evaluated against members of the hall of fame.

A special type of hall of fame is a Pareto archive, so you may also search for this term.

Look up papers on these issues, find one that you think you can implement in MM-NEAT, and then take a shot at implementing it. Several extra classes and maybe even some packages will probably be needed, so let's brainstorm together after you find a paper to work from.

Additionally, even though we are primarily interested in applying this technique to board game co-evolution, the code should be written in a general way so that the technique can be applied to any domain that uses single-population coevolution ... for example, this code might eventually be used for the microRTS task, or some other future task.

**Comment by schrum2** [comment]
Here are links to several papers about hall of fames (put the in Google Drive):
https://www.lri.fr/~hansen/proceedings/2013/GECCO/companion/p1683.pdf
http://people.idsia.ch/~tom/publications/scalable.pdf
http://www.lcc.uma.es/~ccottap/papers/nogueira13halloffame.pdf
http://nn.cs.utexas.edu/downloads/papers/lubberts.coevolution-gecco01.pdf

One of them is actually about a Hall of Fame used in an RTS game ( @alicequint ), which brings up the point that allowing for co-evolution with a hall of fame is a general approach that can be applied to microRTS as well. Most of the hall of fame code changes will actually be associated with the SinglePopulationCoevolutionTask that you both use.

**Comment by schrum2** [comment]
Tackling this issue will probably require the creation of a new class HallOfFame that somehow stores genotypes and is somehow used by single population coevolution, but give a sketch of an implementation idea you have (or ask questions you need answers to) before diving in.

**Comment by schrum2** [comment]
Although this departs slightly from how hall of fames are implemented elsewhere, I think the following is a good first approach for how to implement the hall of fame.It will also allow us to try a variety of different options in the future.

Make a HallOfFameFitness that makes the agent play the game against all members of the hall of fame, in a manner similar to the static opponent other score. Later, we can tweak this to play against only a portion of the hall of fame in various ways.

Note that I would actually like have the option to use this fitness score in addition to the standard win/draw/loss fitness. In fact, always having only one fitness function is pretty limiting. The board game code should be generalized to support multiple fitness functions rather than always assuming that there will be only one fitness function at the start of the list followed by other scores. Instead, maintain a variable number of fitness functions and a variable number of other scores. 

This will require some big code changed, but is the better design approach long term. You should probably generalize the code to support multiple fitness functions first, and then worry about the HallOfFameFitness next.

**Comment by schrum2** [comment]
Correction: when you get around to writing the HallOfFameFitness, this will be specific to board games. However, you should set up most of the code in a general fashion that will work for all domains. To do this, you will need a method(s) in the HallOfFame class that takes as input a genotype (to evaluate) and some way of specifying a portion of the HallOfFame to evaluate against. This method will access MMNEAT.task and cast it to a single pop coevolution task in order to evaluate the genotype against genotypes from the hall of fame, and then return average fitness(es).

Still, you need to support multiple fitness functions in board games first.

**Comment by schrum2** [comment]
Make a batch file and run a hall of fame experiment over the weekend if you can.

**Comment by DarwinJohnson** [comment]
Attempted to run the Checkers HallOfFame Batch File, but it crashed. Attempted to run the Othello HallOfFame Batch File, but it had the same crash. Documented the Crash with a picture. Looking at the code, I'm not sure what happened. As far as I can tell, the HallOfFame in the SinglePopulationCompetativeCoevolutionTask isn't being initialized early enough for the HallOfFame Fitness Function to access it, but that may not be the case. Here is the crash log:

![othello halloffame crash](https://user-images.githubusercontent.com/27302097/27499330-e14effe8-5828-11e7-994e-a468b60315e5.png)

**Comment by schrum2** [comment]
I think you are right, and the error makes sense. At the point when the task is being constructed, it is trying to construct the Hall of fame fitness before the task is available in MMNEAT.task.

This is similar to the issue with the boardGame variable, The solution is the same. I guess you'll need to move the hall of fame variable into MMNEAT after all, and initialize it before the task is constructed.

**Comment by DarwinJohnson** [comment]
Running the Othello and Checkers Hall Of Fame Batch Files now. Also running the Non-HallOfFame counterparts to provide something to compare against. Will document the results after the weekend.

**Comment by DarwinJohnson** [comment]
Tried to run both Hall Of Fame Batch Files over the weekend, but they appear to have crashed on the first Generation. Both files have the same crash log; there is a recurring set of errors. I have documented the crash and highlighted one iteration of the errors.

![halloffame crash](https://user-images.githubusercontent.com/27302097/27543299-76229724-5a4f-11e7-8dc6-d068e6c42511.png)

**Comment by DarwinJohnson** [comment]
In addition to the above error, there appears to have been a problem with keeping track of Fitness Scores for the regular SinglePop Coevolution files. Looking at Checkers in particular, it looks like the Advanced Fitness and the Other Opponent Fitness stayed at 0 for all 500 Generations. I think I figured out the problem, and I'm running another Batch File to see if this error has been fixed. Maybe it's related to the above problem?
